### PR TITLE
Add MUI demo

### PR DESCRIPTION
**Pull request by h-contributeur**
TL;DR: This aims at providing an example using Google Material UI using [MUI](https://mui.com/core/) and [Formik](https://formik.org/), in order to solve [this issue](https://github.com/enuchi/React-Google-Apps-Script/issues/141).

The issue concerning the bundle size is addressed by updating `tsconfig.json` and `webpack.config.js`, to enable tree-shaking properly, even in Typescript. A new file `global.d.ts` is also added to keep functions at the root.

Details:
- `tsconfig.json` now includes `"moduleResolution": "node"` and `"target": "ES2019"`.
- `webpack.config.js` now uses `ecma: 6`
- a new file `global.d.ts` allows to keep functions at the root in `main.gs`
- a new dialog is included: `dialog-demo-mui`, with the necessary function to show it in the menu, and to open the dialog.
- Formik is added for MUI demo dialog.

Final considerations:
The bundle size is still big comparatively to other dialogs: **200kb versus 10kb** (including Formik 40kb).
This is partly due to:
- MUI not being included via CDN, as [it is not recommended ](https://mui.com/material-ui/getting-started/installation/#cdn). This would in fact download the whole library. So, the comparison is unfair since other dialogs do not include their respective packages in their HTML file.
- Formik is not included via CDN as I haven't found how to make it work. It seems that they have issues with CDN, as they have several issues about it.

I have tested including a heavy MUI component to see if the size of the HTML changes. Including [Autocomplete](https://mui.com/material-ui/react-autocomplete/#combo-box) adds 60kb to the bundle size. This is the proof that tree-shaking is in place and working.

### CODE REVIEW

**Comment by enuchi** [comment]
Thanks @h-contributeur , really appreciate the PR. Have a few thoughts on these changes, but first I wanted to check if I understand the bundle sizes correctly. It's 200kb total to include MUI with tree-shaking (including about 40kb for Formik), compared with around 10kb without the libraries.

MUI's production minified package (entire library) from CDN at https://unpkg.com/@mui/material@5.11.11/umd/material-ui.production.min.js is about 140kb. I know they recommend against loading from a CDN etc. in typical situations (I'd argue this isn't the case here with Google Apps Script) but is this right that the full minified version is still smaller than the bundle created with tree-shaking?

**Comment by enuchi** [comment]
Thanks @h-contributeur. That would be great. I played around with your PR and found I could use a CDN to load MUI with a few changes. Could you incorporate these into your PR?

I'd like to keep the "named export" version of global assignment for the server code, since I think adding back the `global.function = ...` syntax is too confusing for most users. Could you revert the tsconfig changes? Alternatively I think this could be improved by having a separate tsconfig.json for server folder with classic moduleResolution that is compatible with gas-webpack-plugin, and one for the client folder that uses more modern module resolution but haven't look much into this.

Lastly, would you mind removing Formik? Think it is a great tool for apps with a lot of forms but not relevant for most here and think it's more useful to just have a plain example showing MUI.

**Comment by enuchi** [comment]
Something like this but with `packageVersion` in the urls. The DynamicCdnWebpackPlugin will not work correctly unless peer dependencies are included.

https://github.com/enuchi/React-Google-Apps-Script/blob/0b7bdd18c0cd3372fbf62652ab297ac156fc2c08/webpack.config.js#L202-L222

And imports like this:
https://github.com/enuchi/React-Google-Apps-Script/blob/0b7bdd18c0cd3372fbf62652ab297ac156fc2c08/src/client/dialog-demo-mui/components/FormInput.tsx#L2

File: src/client/dialog-demo-mui/styles.css [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/* needed to make consistent test snapshots across OSs */
+body {
+  font-family: Arial !important;
+}
+
+/*
+CSSTransitionGroup styling
```
Comment (enuchi) [inline src/client/dialog-demo-mui/styles.css:7]: we probably don't need these since you didn't use react-transition-group in the MUI example

File: webpack.config.js [hunk 2]
```diff
@@ -207,6 +213,27 @@ const DynamicCdnWebpackPluginConfig = {
           version: packageVersion,
           url: `https://unpkg.com/react-bootstrap@${packageVersion}/dist/react-bootstrap${packageSuffix}`,
         };
+      case '@mui/material':
+        return {
+          name: packageName,
+          var: 'MaterialUI',
+          version: packageVersion,
+          url: `https://unpkg.com/@mui/material@${packageVersion}/umd/material-ui.production${packageSuffix}`,
```
Comment (enuchi) [inline webpack.config.js:221]: Looks like they use a different naming convention here, so should be like:

`https://unpkg.com/@mui/material@${packageVersion}/umd/material-ui.${
            isProd ? 'production.min.js' : 'development.js'
          }`

Comment (enuchi) [reply webpack.config.js:221]:
```suggestion
          url: `https://unpkg.com/@mui/material@${packageVersion}/umd/material-ui.${ isProd ? 'production.min.js' : 'development.js' }`,
```

File: webpack.config.js [hunk 3]
```diff
@@ -207,6 +213,27 @@ const DynamicCdnWebpackPluginConfig = {
...
+          var: 'emotionReact',
+          version: packageVersion,
+          url: `https://unpkg.com/@emotion/react@${packageVersion}/dist/emotion-react.umd${packageSuffix}`,
+        };
+      case '@emotion/styled':
+        return {
+          name: packageName,
+          var: 'emotionStyled',
+          version: packageVersion,
+          url: `https://unpkg.com/@emotion/styled@${packageVersion}/dist/emotion-styled.umd${packageSuffix}`,
```
Comment (enuchi) [inline webpack.config.js:235]:
```suggestion
          url: `https://unpkg.com/@emotion/styled@${packageVersion}/dist/emotion-styled.umd.min.js`,
```

**Review by enuchi — COMMENTED** [review]
Do you mind making a few minor changes?

**Comment by enuchi** [comment]
Looks good! Adding MUI is increasing the overall bundle size here during development and making the tests fail here, so am going to make a couple changes to the repo. Do you mind pulling in some changes once they're in?

**Comment by enuchi** [comment]
Ok was able to update the PR -- looks good now, going to merge this.

**Comment by enuchi** [comment]
Thanks for your help getting this example into the project @h-contributeur.
